src/server/command/worldedit/pattern.zig

```diff
@@ -0,0 +1,58 @@
+const std = @import("std");
+
+const main = @import("root");
+const AliasTable = main.utils.AliasTable;
+const Block = main.blocks.Block;
+const ListUnmanaged = main.ListUnmanaged;
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+
+blocks: AliasTable(Entry),
+
+const Entry = struct {
+	block: Block,
+	chance: f32,
+};
+
+pub fn initFromString(allocator: NeverFailingAllocator, source: []const u8) !@This() {
+	var specifiers = std.mem.splitScalar(u8, source, ',');
+	var totalWeight: usize = 0;
+
+	var weightedEntries: ListUnmanaged(struct {block: Block, weight: usize}) = .{};
+	defer weightedEntries.deinit(main.stackAllocator);
+
+	while(specifiers.next()) |specifier| {
+		var iterator = std.mem.splitScalar(u8, specifier, '%');
+
+		const first = iterator.next();
+		const second = iterator.next();
```

Comment:
You could also use `.first()` and `.rest()`. I think that would make the error handling below easier, since only one value could be null. All the other errors would then be automatically catched as invalid ids.
Same for the `peek` check.

Reply:
Umm if you use first and rest then neither is null. Both return empty string if there is nothing to split.

Reply:
Hmm. Why not just use indexOf instead then?

Reply:
Mind that `parseBlock` never fails.

Reply:
Yeah, a poor design decision. But this should not be your concern here.